### PR TITLE
Add backup extension resource documentation

### PR DESCRIPTION
**Pull request by swapnilgm**
Signed-off-by: Swapnil Mhamane <swapnil.mhamane@sap.com>

**What this PR does / why we need it**:
This PR adds the documentation for extension BackupBucket and BackupEntry resources.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: docs/extensions/backupbucket.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+## What is the scope of bucket?
+
+A bucket will be provisioned per `Seed`. So, backup of every `Shoot` created on that `Seed` will be stored under different shoot specific prefix under the bucket.
+For the backup of the `Shoot` rescheduled on different `Seed` it will continue to use the same bucket.
+
+## What is the lifespan of `BackupBucket`?
+The bucket associated with `BackupBucket` will be created at creation of `Seed`. And as per current implementation, it will be deleted on deletion of `Seed` and there isn't any `BackupEntry` resource associated with it.
+
+Eventually, in future the the deletion logic will be changed such that on deletion of `Seed` if there isn't anu schedulable `Seed` available and there isn't any associated `BackupEntry` resource.
```
Comment (zanetworker) [inline docs/extensions/backupbucket.md]:
```suggestion
Eventually, in the future, the deletion logic will be changed such that on deletion of a `Seed` if there isn't any schedulable `Seed` available and there isn't any associated `BackupEntry` resource.
```

Comment (zanetworker) [reply docs/extensions/backupbucket.md]: I also think this sentence is missing something.

File: docs/extensions/backupbucket.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: backupprovider
+    namespace: shoot--foo--bar
+```
+
+The `.spec.secretRef` contains a reference to the provider secret pointing to the account that shall be used to create the needed resources. This provider secret will be configured
+by Gardener operator in `seed` resource and propagated over here by seed controller.
```
Comment (zanetworker) [inline docs/extensions/backupbucket.md]:
```suggestion
by Gardener operator in the `seed` resource and propagated over there by seed controller.
```

File: docs/extensions/backupbucket.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+After your controller has created the required bucket, if required it needs generate the secret to access the objects in buckets and put reference to it in `status`. This secret is
```
Comment (zanetworker) [inline docs/extensions/backupbucket.md]:
```suggestion
After your controller has created the required bucket, if required it generates the secret to access the objects in buckets and put reference to it in `status`. This secret is
```

File: docs/extensions/backupentry.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Contract: `BackupEntry` resource
+
+The Gardener project features a sub-project called [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) to take periodic backups of etcd backing Shoot clusters. It demands the bucket (or its equivalent in different object store providers) access credentials to be created and configured externally with appropriate credentials. The `BackupEntry` resource takes this responsibility in Gardener to provide this information by creating secret specific to the component. Said that, core motivation for introducing this resource was to support retention  of backups post deletion of `Shoot`. The etcd-backup-restore components takes responsibility of garbage collecting old backups out of define period. Once shoot is deleted, we need to persist the backups for few days. Hence, Gardener uses the BackupEntry resource for this housekeeping work post deletion of `Shoot`. The `BackupEntry` resource is responsible for shoot specific prefix under referred bucket.
```
Comment (zanetworker) [inline docs/extensions/backupentry.md]:
```suggestion
The Gardener project features a sub-project called [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) to take periodic backups of etcd backing Shoot clusters. It demands the bucket (or its equivalent in different object store providers) access credentials to be created and configured externally with appropriate credentials. The `BackupEntry` resource takes this responsibility in Gardener to provide this information by creating a secret specific to the component. Said that, the core motivation for introducing this resource was to support retention  of backups post deletion of `Shoot`. The etcd-backup-restore components takes responsibility of garbage collecting old backups out of the defined period. Once a shoot is deleted, we need to persist the backups for few days. Hence, Gardener uses the `BackupEntry` resource for this housekeeping work post deletion of a `Shoot`. The `BackupEntry` resource is responsible for shoot specific prefix under referred bucket.
```

File: docs/extensions/backupentry.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+Before introducing the `BackupEntry` extension resource Gardener was using Terraform in order to create and manage these provider-specific resources (e.g., see [here](https://github.com/gardener/gardener/tree/0.27.0/charts/seed-terraformer/charts/aws-backup)).
+Now, Gardener commissions an external, provider-specific controller to take over this task. You can also refer to backupInfra proposal documentation to get idea about how the transition was done and understand the resource in broader scope.
+
+## What is the lifespan of `BackupEntry`?
+The bucket associated with `BackupEntry` will be created at using `BackupBucket` resource. The `BackupEntry` resource will be created as a part of `Shoot` creation. But resource continue to exists post deletion of `Shoot`. The  `deletionGracePeriod` of `BackupEntry` resource i.e. time after the shoot deletion
```
Comment (zanetworker) [inline docs/extensions/backupentry.md]:
```suggestion
The bucket associated with `BackupEntry` will be created at using `BackupBucket` resource. The `BackupEntry` resource will be created as a part of a `Shoot` creation. But resource continue to exists post deletion of a `Shoot`. The  `deletionGracePeriod` of a `BackupEntry` resource i.e. time after the shoot deletion
```

File: docs/extensions/backupentry.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  region: eu-west-1
+  bucketName: foo
+  secretRef:
+    name: backupprovider
+    namespace: shoot--foo--bar
+```
+
+The `.spec.secretRef` contains a reference to the provider secret pointing to the account that shall be used to create the needed resources. This provider secret will be propagated from `BackupBucket` resource by Shoot controller.
+
+Your controller is supposed to create the `etcd-backup` secret in control-plane namespace for shoot. This secret is supposed to be used by Gardener or eventually etcd-backup-restore component to backup the etcd. The controller implementation should cleanup the objects created under shoot specific prefix in bucket equivalent to name of `BackupEntry` resource.
```
Comment (zanetworker) [inline docs/extensions/backupentry.md]:
```suggestion
Your controller is supposed to create the `etcd-backup` secret in control-plane namespace of a shoot. This secret is supposed to be used by Gardener or eventually the etcd-backup-restore component to backup the etcd. The controller implementation should cleanup the objects created under shoot specific prefix in bucket equivalent to name of `BackupEntry` resource.
```

File: docs/extensions/backupbucket.md (hunk 1 above)
Comment (swapnilgm) [reply docs/extensions/backupbucket.md]: Thank you for the feedback. I will try to rephrase it.

File: docs/extensions/backupentry.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+In order to support a new infrastructure provider you need to write a controller that watches all `BackupBucket`s with `.spec.type=<my-provider-name>`. You can take a look at the below referenced example implementation for the Azure provider.
+
+## References and additional resources
+
+* [`BackupEntry` API Reference](https://gardener.cloud/api-reference/extensions/#extensions.gardener.cloud/v1alpha1.BackupBucket)
+* [Exemplary implementation for the Azure provider](https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-azure/pkg/controller/backupentry)
+* [`BackupBucket` resource documentation](./backupbucket.md)
+* [Shared bucket proposal](../proposals/02-backupinfra.md)
+* [Gardener-controller-manager-component-config API specification](https://github.com/gardener/gardener/blob/master/pkg/controllermanager/apis/config/types.go)
```
Comment (zanetworker) [inline docs/extensions/backupentry.md]: maybe also provide the permalink for the backup configuration (referring to the related lines in the code).

Comment (swapnilgm) [reply docs/extensions/backupentry.md]: Done

**Comment by swapnilgm** [comment]
@zanetworker Thank you for quick review. I've addressed the suggestions. PTAL.

**Review by zanetworker — APPROVED** [review]
/lgtm

**Pull request merged**